### PR TITLE
Possibility to negate script filter. Fixes #577

### LINKED ISSUE #577: Enhancement: Add the possibility to negate script filter
Hello,

I have been using Ivre for a while and thought it would be interesting to have the ability to do `!script:id` and `!script:id:output`. Sometimes it is interesting to know why a particular script didn't run or identify the hosts for which the output is different. It would also allow to identify "blank" results when the host timeout didn't allow nmap to store any result so they don't have any of the host scripts.

I'm ready to implement it myself. Just wondering if adding a "not" clause in the `searchscan` (and adding it to the GUI/CLI) would do the trick or if things might get more complicated.

Thanks.

### PR DESCRIPTION
**Pull request by Golbark**
Fixes #577

### CODE REVIEW

**Comment by p-l-** [comment]
Thanks for this PR. I have a couple of remarks:
* You are updating `.searchscript()` for MongoDB, but you should also update it in PostgreSQL (and update the method prototype in `db/__init__.py`).
* You should add some tests to your contribution, to help us prevent regressions.

For some reason, Travis-CI did not run the regression tests on this PR. I will close it and re-open it in the hope it will fix this problem.

**Comment by p-l-** [comment]
Yes, the tests should be added in `tests.py`. Probably both in the "view" and "nmap" tests.

Regarding PostgreSQL and the values, actually you cannot assume that it is only called by other filters, since people may use IVRE's API in their own projects (`.searchscript()` is a public API).

**Comment by p-l-** [comment]
I have contacted Travis-CI support, they have been very helpful and are working on the issue.

File: ivre/db/sql/__init__.py [hunk 1]
```diff
@@ -1331,12 +1333,32 @@ def _to_json(key, value):
...
+                                 cls.tables.port)
+                        )
+                    )
+                    .where(req))
             return cls.base_filter(script=[(
                 req,
-                [func.jsonb_array_elements(cls.tables.script.data[subkey])
-                 .alias(subkey.replace('.', '_').replace('-', '_'))
-                 for subkey in needunwind],
+                [func.jsonb_array_elements(cls.tables.script.data[subkey2])
```
[the remark this answers is not in the document]
Comment (Golbark) [reply ivre/db/sql/__init__.py]: I had a flake8 error message because subkey overwrites another variable named subkey above

**Comment by p-l-** [comment]
I'll have a look. BTW nice catch for the SSL test! Would you mind making a distinct PR for that?

**Comment by Golbark** [comment]
Hi @p-l- ,

Not sure to understand you? The ssl-cert test was added by me in the sole purpose of testing this new no-script option but apparently it's failing for postgresql only.

The test is simple enough: count(script(x)) + count(no-script(x)) = count(*).

I have no idea why it's failing for postgresql. Either the view_count on which I'm relying is computed incorrectly or my query is wrong. However, the nmap test succeed and the `searchscript` method is only defined once. I'll keep investigate but any hint is welcome :)

**Comment by p-l-** [comment]
@Golbark my bad, I though your commit "Fix typo in cgi filter tests" was fixing an existing mistake in IVRE :). I'll see what I can do to help you!

**Comment by Golbark** [comment]
I have found the issue. I took the assumption that all scan objects have at least one port attached to them which is true for nmap DB but not for passive DB. As a result, my query does not include those scans without ports when selecting the no-script xxx filter. 

The easiest way to solve this issue and make the code more simple is to change the (...) WHERE EXISTS (...) in (...) WHERE NOT EXISTS (...) in the function `query()` of `ActiveFilter`. I'm thinking about using similar methodology as for the port or the hostname by including a boolean as first element of the list to indicate whether it needs to be included or excluded.

**Comment by p-l-** [comment]
A little comment here: it is, actually very possible to have a scan result in Nmap that has no port associated with it. Maybe it would be useful to add a sample to test everything works with it.

**Comment by p-l-** [comment]
Thanks a lot for this contribution and for your perseverance! You're totally right about the Codacy issue, I don't understand it.

Thanks again!

**Comment by Golbark** [comment]
Thanks ! It was not as simple as I first thought but at least I got to learn about the inner mechanisms of IVRE :)

**Comment by p-l-** [comment]
I hope to see you again soon for more contributions then :)

**Pull request merged**